### PR TITLE
TandemFoil: golden config + no-EMA + model capacity 4L/256d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The default model (3L/192d/3H) may be capacity-limited at 11 epochs. A larger model (4L/256d/4H) has ~78% more parameters, enabling richer representations of the pressure field. At slices=64, the larger model may still complete 8-10 epochs within the 30-min timeout — enough to benefit from the added capacity. AirfRANS 4L/256d tests were at only 2 epochs with Fourier; this is the first test of 4L/256d on TandemFoil with many epochs available.

## Instructions

Run sequentially (ONE AT A TIME). Golden config: slices=64, T_max=30, no-EMA, Lion lr=3e-4.

```bash
cd target/icml2026

# Run 1: 4L/256d/4H, golden config, no-EMA
python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 \
  --cosine_t_max 30 --no-use-ema --model_slices 64 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --wandb_group kaneda-golden-capacity --wandb_name kaneda-golden-4l256d

# Run 2: 5L/320d/5H, golden config, no-EMA (if time permits — only after Run 1 completes)
python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 \
  --cosine_t_max 30 --no-use-ema --model_slices 64 \
  --model-layers 5 --model-hidden-dim 320 --model-heads 5 \
  --wandb_group kaneda-golden-capacity --wandb_name kaneda-golden-5l320d
```

**Report** `val_primary/surface_pressure_mae`, test breakdown by split, **epochs completed** (critical — note if larger model reduces epoch count), and loss trajectory. Target: beat **114.92**.

## Baseline

| Dataset | Metric | Value | PR |
|---|---|---|---|
| TandemFoil | val_primary/surface_pressure_mae | **114.92** | #2435 (slices=64, T_max=30, Lion lr=3e-4, EMA=True, 11 epochs, 3L/192d) |
| TandemFoil | test_primary/surface_pressure_mae | **108.16** | #2435 |

Reproduce baseline: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine_t_max 30 --model_slices 64`

**Note:** If 4L/256d gets fewer than 8 epochs, the capacity gain may not outweigh the epoch loss. Report epoch count carefully.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kaneda:

## Results

### Run 1: 4L/256d/4H — golden config + no-EMA

| Metric | Baseline (3L/192d) | 4L/256d/4H | Delta |
|---|---|---|---|
| **val_primary/surface_pressure_mae** | **114.92** | 224.40 | +95.3% ❌ |
| **test_primary/surface_pressure_mae** | **108.16** | 216.63 | +100.3% ❌ |
| test_single_in_dist | — | 224.10 | — |
| test_geom_camber_rc | — | 189.00 | — |
| test_geom_camber_cruise | — | 226.00 | — |
| test_re_rand | — | 227.43 | — |
| **Epochs completed** | **11** | **2** | -82% |

- **W&B run:** 2qi6a8tv
- **Peak VRAM:** ~47 GB
- **Command:** `CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 30 --no-use-ema --model-slices 64 --model-layers 4 --model-hidden-dim 256 --model-heads 4 --wandb-group kaneda-golden-capacity --wandb-name kaneda-golden-4l256d --agent kaneda`

### Run 2: 5L/320d/5H — golden config + no-EMA

| Metric | Baseline (3L/192d) | 5L/320d/5H | Delta |
|---|---|---|---|
| **val_primary/surface_pressure_mae** | **114.92** | 206.50 | +79.7% ❌ |
| **test_primary/surface_pressure_mae** | **108.16** | 192.62 | +78.1% ❌ |
| test_single_in_dist | — | 217.36 | — |
| test_geom_camber_rc | — | 191.84 | — |
| test_geom_camber_cruise | — | 174.67 | — |
| test_re_rand | — | 186.63 | — |
| **Epochs completed** | **11** | **2** | -82% |

- **W&B run:** fwilsngh
- **Peak VRAM:** ~47 GB
- **Command:** `CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 30 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 320 --model-heads 5 --wandb-group kaneda-golden-capacity --wandb-name kaneda-golden-5l320d --agent kaneda`

### What happened

**Both runs failed to beat the baseline — epoch starvation is the bottleneck.** The larger models take ~15 min/epoch at slices=64 vs ~2.7 min/epoch for 3L/192d, resulting in only 2 epochs in the 30-min timeout vs 11 for the baseline. The capacity gain cannot overcome the 5.5x training time reduction.

Interesting observation: **5L/320d actually outperformed 4L/256d at the same epoch count** (val 206.50 vs 224.40, test 192.62 vs 216.63). This suggests larger models learn faster per-epoch, but under time-constrained training the epoch count matters far more.

The loss trajectory was steep and still improving at cutoff (epoch 1→2: loss 1.57→0.80 for 5L, 1.51→0.81 for 4L), indicating these models would likely catch up given enough epochs — but they need a 5-10x longer timeout or a way to reduce per-epoch cost.

### Suggested follow-ups

1. **Intermediate capacity (3L/256d or 4L/192d):** Test models that are slightly larger than 3L/192d but don't blow up per-epoch time. 3L/256d would add width without depth cost; 4L/192d would add depth cheaply.
2. **Gradient accumulation + smaller batch:** If per-step memory is the bottleneck, smaller batches with accumulation could allow larger models to train faster per epoch.
3. **Mixed precision / compile:** torch.compile or AMP could reduce per-epoch time enough to recover epochs for the larger models.
4. **slices=48 with 4L/256d:** Reducing slices further might recover enough epochs to make the capacity gain worthwhile.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2488 because Dead end under the current timeout constraints. Both 4L/256d (224.40) and 5L/320d (206.50) were severely epoch-starved — only 2 epochs vs the baseline's 14 epochs. The per-epoch cost at these model sizes (~15 min/epoch) consumes the entire budget. The failure mode is structural, not ML-related — the hypothesis remains interesting.

Follow-up direction: The interesting intermediate configs are 3L/256d (width-only increase, preserving depth) or 4L/192d. These should maintain reasonable epoch counts while expanding capacity. With the new 180-min budget now available, a fresh test of 3L/256d or 4L/256d at slices=64 with --epochs 999 is worth assigning. Note: always pass --epochs 999 explicitly to avoid the epochs=2 default bug.

The TandemFoil baseline is now 82.65 (PR #2473, Fourier+physics+no-EMA). Assigning you a new experiment soon..
